From 6d166b2fa7cd6e8d47eb027f473a6315eebb66f6 Mon Sep 17 00:00:00 2001
From: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
Date: Wed, 15 May 2024 11:25:19 +0800
Subject: [PATCH 057/107] mtk: mac80211: add mlo probe client support

Add link unspec for mld case; otherwise it would be link 0
instead of primary link

Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>

Send the null func in min rate.
Without this patch, the null data frame will fail to tx in both legacy
AP and MLD AP.
(appears in ICS log but not in air).

Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
---
 net/mac80211/cfg.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/net/mac80211/cfg.c b/net/mac80211/cfg.c
index dd2c3300..76c55548 100644
--- a/net/mac80211/cfg.c
+++ b/net/mac80211/cfg.c
@@ -4338,7 +4338,10 @@ static int ieee80211_probe_client(struct wiphy *wiphy, struct net_device *dev,
 	info = IEEE80211_SKB_CB(skb);
 
 	info->flags |= IEEE80211_TX_CTL_REQ_TX_STATUS |
-		       IEEE80211_TX_INTFL_NL80211_FRAME_TX;
+		       IEEE80211_TX_INTFL_NL80211_FRAME_TX |
+		       IEEE80211_TX_CTL_USE_MINRATE;
+	if (ieee80211_vif_is_mld(&sdata->vif))
+		info->control.flags |= IEEE80211_TX_CTRL_MLO_LINK_UNSPEC;
 	info->band = band;
 
 	skb_set_queue_mapping(skb, IEEE80211_AC_VO);
-- 
2.45.2

